### ISSUE TITLE
ADO.Net connection pooling doesn't seem to support multiple threads well

### ISSUE THREAD

**Issue by joshmouch** [dotnet/core#4221]
I'm trying to spin up 50 threads to immediately open 50 `SqlConnection`s to a Microsoft SQL Server connection pool.  However, no matter what I try, the threads seem to only open one at a time.  In other words, if each connection takes 2 seconds to open, then the 50'th thread's `Open` method will take 100 seconds to complete.  Further to this point, if the Connection Timeout isn't set to > 100 seconds then the connection will time out.  

It's like there's something blocking internally to only allow this one-at-a-time addition to the connection pool.
I've asked around on stack overflow, and the responses seem to suggest that if you open 50 connections in parallel, they should all open.

So this makes me wonder if this is a .Net Core bug or "feature".

**Comment by benaadams** [comment]
How are you spinning up the threads?

**Comment by joshmouch** [comment]
@benaadams ,  I've tried this two ways: with and without threads (without threads, I just set "Min Pool Size=50" in the connection string then open a single connection, which triggers some background process to start opening connections.  

But one of the iterations with threads is:

```

			var tasks = new List<Task>();
			var connections = new ConcurrentBag<SqlConnection>();
			for (int i = 0; i < 50; i++)
			{
				var task = new Task(() =>
				{
					var connection = new SqlConnection(sqlConnectionString);
					connections.Add(connection);
					connection.Open();
					var a = 1; // If I set breakpoint here, there is a 2 second delay between each time it gets hit.  I get a Connection Timeout before the 16th iteration.
				}, TaskCreationOptions.LongRunning);
				task.Start();
				tasks.Add(task);
			}

			Task.WaitAll(tasks.ToArray());

			tasks = new List<Task>();
			foreach (var connection in connections)
			{
				var task = new Task(() =>
				{
					connection.Close();
				}, TaskCreationOptions.LongRunning);
				task.Start();
				tasks.Add(task);
			}
			Task.WaitAll(tasks.ToArray());
```

**Comment by joshmouch** [comment]
I should also add that I'm running the following SQL to monitor the connections, and I see them getting created at a rate of 1 per 2 seconds using both methods (multithreading or a single thread with Min Pool Size set).

```
SELECT  des.program_name
      , des.login_name
      , des.host_name
      , COUNT(des.session_id) [Connections]
FROM    sys.dm_exec_sessions des
INNER JOIN sys.dm_exec_connections DEC
        ON des.session_id = DEC.session_id
WHERE   des.is_user_process = 1
        AND des.status != 'running'
GROUP BY des.program_name
      , des.login_name
      , des.host_name
HAVING  COUNT(des.session_id) > 2
ORDER BY COUNT(des.session_id) DESC
```

**Comment by joshmouch** [comment]
In the multi-threading scenario, if I look at the thread list once all 50 threads start, most of them (48ish) are waiting with the following stack track:

```
Not Flagged		44212	41	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple(System.ReadOnlySpan<System.Threading.WaitHandle> waitHandles, bool waitAll, int millisecondsTimeout)
System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple(System.Threading.WaitHandle[] waitHandles, bool waitAll, int millisecondsTimeout)
System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitAny(System.Threading.WaitHandle[] waitHandles, int millisecondsTimeout)
Microsoft.Data.SqlClient.dll!Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(System.Data.Common.DbConnection owningObject, uint waitForMultipleObjectsTimeout, bool allowCreate, bool onlyOneCheckConnection, Microsoft.Data.Common.DbConnectionOptions userOptions, out Microsoft.Data.ProviderBase.DbConnectionInternal connection)
Microsoft.Data.SqlClient.dll!Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(System.Data.Common.DbConnection owningObject, System.Threading.Tasks.TaskCompletionSource<Microsoft.Data.ProviderBase.DbConnectionInternal> retry, Microsoft.Data.Common.DbConnectionOptions userOptions, out Microsoft.Data.ProviderBase.DbConnectionInternal connection)
Microsoft.Data.SqlClient.dll!Microsoft.Data.ProviderBase.DbConnectionFactory.TryGetConnection(System.Data.Common.DbConnection owningConnection, System.Threading.Tasks.TaskCompletionSource<Microsoft.Data.ProviderBase.DbConnectionInternal> retry, Microsoft.Data.Common.DbConnectionOptions userOptions, Microsoft.Data.ProviderBase.DbConnectionInternal oldConnection, out Microsoft.Data.ProviderBase.DbConnectionInternal connection)
Microsoft.Data.SqlClient.dll!Microsoft.Data.ProviderBase.DbConnectionInternal.TryOpenConnectionInternal(System.Data.Common.DbConnection outerConnection, Microsoft.Data.ProviderBase.DbConnectionFactory connectionFactory, System.Threading.Tasks.TaskCompletionSource<Microsoft.Data.ProviderBase.DbConnectionInternal> retry, Microsoft.Data.Common.DbConnectionOptions userOptions)
Microsoft.Data.SqlClient.dll!Microsoft.Data.ProviderBase.DbConnectionClosed.TryOpenConnection(System.Data.Common.DbConnection outerConnection, Microsoft.Data.ProviderBase.DbConnectionFactory connectionFactory, System.Threading.Tasks.TaskCompletionSource<Microsoft.Data.ProviderBase.DbConnectionInternal> retry, Microsoft.Data.Common.DbConnectionOptions userOptions)
Microsoft.Data.SqlClient.dll!Microsoft.Data.SqlClient.SqlConnection.TryOpen(System.Threading.Tasks.TaskCompletionSource<Microsoft.Data.ProviderBase.DbConnectionInternal> retry)
Microsoft.Data.SqlClient.dll!Microsoft.Data.SqlClient.SqlConnection.Open()


```

**Comment by joshmouch** [comment]
If I use `OpenAsync` instead of `Open`, the stuck thread has a similar stack trace:
```
Not Flagged		62476	26	Worker Thread	<No Name>	System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple
System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple(System.ReadOnlySpan<System.Threading.WaitHandle> waitHandles, bool waitAll, int millisecondsTimeout)
System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitMultiple(System.Threading.WaitHandle[] waitHandles, bool waitAll, int millisecondsTimeout)
System.Private.CoreLib.dll!System.Threading.WaitHandle.WaitAny(System.Threading.WaitHandle[] waitHandles, int millisecondsTimeout)
Microsoft.Data.SqlClient.dll!Microsoft.Data.ProviderBase.DbConnectionPool.TryGetConnection(System.Data.Common.DbConnection owningObject, uint waitForMultipleObjectsTimeout, bool allowCreate, bool onlyOneCheckConnection, Microsoft.Data.Common.DbConnectionOptions userOptions, out Microsoft.Data.ProviderBase.DbConnectionInternal connection)
Microsoft.Data.SqlClient.dll!Microsoft.Data.ProviderBase.DbConnectionPool.WaitForPendingOpen()
System.Private.CoreLib.dll!System.Threading.ThreadHelper.ThreadStart_Context(object state)
System.Private.CoreLib.dll!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, object state)
System.Private.CoreLib.dll!System.Threading.ThreadHelper.ThreadStart()

```

**Comment by benaadams** [comment]
I think unless you are specifying `MaxPoolSize=N` in the connection string it should default to 100 connections; so is a bit strange. /cc @cheenamalhotra 

You might want to open this issue in https://github.com/dotnet/runtime (or have it transferred)

**Comment by joshmouch** [comment]
I have MaxPoolSize=500 just in case that was a factor, but it wasn't.

**Comment by joshmouch** [comment]
I should also note that if I turn connection pooling off, these 50 connections open almost all at the same time then close again almost all at the same time.  So the server itself isn't a factor.

**Comment by cheenamalhotra** [comment]
We can transfer this issue to dotnet/sqlclient and investigate more.
